pkg/helm/chartproxy/repos.go

```diff
@@ -0,0 +1,223 @@
+	"k8s.io/client-go/dynamic"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	"github.com/openshift/library-go/pkg/crypto"
+)
+
+var (
+	helmChartRepositoryGVK = schema.GroupVersionResource{
+		Group:    "helm.openshift.io",
+		Version:  "v1beta1",
+		Resource: "helmchartrepositories",
+	}
+)
+
+const (
+	configNamespace = "openshift-config"
+)
+
+type helmRepo struct {
+	Name            string
+	URL             *url.URL
+	TLSClientConfig *tls.Config
+}
+
+type TLSConfigGetter interface {
+	Get() (*tls.Config, error)
+}
+
+func (repo helmRepo) Get() (*tls.Config, error) {
+	return repo.TLSClientConfig, nil
+}
+
+func (repo helmRepo) httpClient() (*http.Client, error) {
+	tlsConfig, err := repo.Get()
+	if err != nil {
+		return nil, err
+	}
+	tr := &http.Transport{
+		TLSClientConfig: tlsConfig,
+	}
+
+	client := &http.Client{Transport: tr}
+	return client, nil
+}
+
+func (hr helmRepo) IndexFile() (*repo.IndexFile, error) {
+	var indexFile repo.IndexFile
+	httpClient, err := hr.httpClient()
+	if err != nil {
+		return nil, err
+	}
+	indexURL := hr.URL.String()
+	if !strings.HasSuffix(indexURL, "/index.yaml") {
+		indexURL += "/index.yaml"
+	}
+	resp, err := httpClient.Get(indexURL)
+	if err != nil {
+		return nil, err
+	}
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	err = yaml.Unmarshal(body, &indexFile)
+	if err != nil {
+		return nil, err
+	}
+	return &indexFile, nil
+}
+
+type HelmRepoGetter interface {
+	List() ([]*helmRepo, error)
+}
+
+type helmRepoGetter struct {
+	Client     dynamic.Interface
+	CoreClient corev1.CoreV1Interface
+}
+
+func (b helmRepoGetter) unmarshallConfig(repo unstructured.Unstructured) (*helmRepo, error) {
+	h := &helmRepo{}
+	urlValue, _, err := unstructured.NestedString(repo.Object, "spec", "connectionConfig", "url")
+	if err != nil {
+		return nil, err
+	}
+
+	h.URL, err = url.Parse(urlValue)
+	if err != nil {
+		return nil, err
+	}
+
+	h.Name, _, err = unstructured.NestedString(repo.Object, "metadata", "name")
+	if err != nil {
+		return nil, err
+	}
+
+	caReference, _, err := unstructured.NestedString(repo.Object, "spec", "connectionConfig", "ca", "name")
+	if err != nil {
+		return nil, err
+	}
+
+	tlsReference, _, err := unstructured.NestedString(repo.Object, "spec", "connectionConfig", "tlsClientConfig", "name")
+	if err != nil {
+		return nil, err
+	}
+
+	var rootCAs *x509.CertPool
+	if caReference != "" {
+		caCert, err := b.configMapValue(caReference, "ca-bundle.crt")
+		if err != nil {
+			return nil, err
+		}
+		if caCert != nil {
+			rootCAs = x509.NewCertPool()
+			if ok := rootCAs.AppendCertsFromPEM(caCert); !ok {
+				return nil, errors.New("Failed to append caCert")
+			}
+		}
+	}
+	if rootCAs == nil {
+		rootCAs, err = x509.SystemCertPool()
+		if err != nil {
+			return nil, err
+		}
+	}
+	h.TLSClientConfig = crypto.SecureTLSConfig(&tls.Config{
+		RootCAs: rootCAs,
+	})
+	if tlsReference != "" {
+		tlsCert, err := b.secretValue(tlsReference, "tls.crt")
```

Comment:
`secretValue()` method is making the API call. Should be enough to call it once and check the wanted field.

```go
        tlsSecret, err := b.CoreClient.Secrets(configNamespace).Get(context.TODO(), name, v1.GetOptions{})
	if err != nil {
		return nil, errors.New(fmt.Sprintf("Failed to GET secret %s", name))

	}
        tlsCert, err := b.secretValue(tlsSecret, "tls.crt")
        tlsCert, err := b.secretValue(tlsSecret, "tls.key")
```